### PR TITLE
REF: Grouper.grouper -> Grouper._gpr_index

### PR DESCRIPTION
**Pull request by jbrockmendel**
ATM we set `self.grouper` in both _get_grouper and in _set_grouper, but to different types.  This changes _set_grouper to set _gpr_index instead, which we can type as `Index | None`

### CODE REVIEW

File: pandas/core/groupby/grouper.py [hunk 1]
```diff
@@ -287,8 +290,11 @@ def __init__(
 
     @final
     @property
-    def ax(self):
-        return self.grouper
+    def ax(self) -> Index:
+        index = self._gpr_index
+        if index is None:
+            raise ValueError("_set_grouper must be called before ax is accessed")
```
Comment (jreback) [inline pandas/core/groupby/grouper.py:296]: this hit in tests?

Comment (jbrockmendel) [reply pandas/core/groupby/grouper.py:296]: no, in practice we always call _set_grouper before accessing this

**Comment by jreback** [comment]
looks fine, pls rebase

**Comment by jbrockmendel** [comment]
rebased + green

**Pull request merged**